Switch internal CUDA build to C++14

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#26757 Switch internal CUDA build to C++14**
* #26522 Remove GCC4 from CI

This doesn't switch any open source builds or CI.
The internal fbcode build is C++17 already for quite some time, but in CUDA code, we had it restricted to C++11.
This diff changes that to C++14.

Because this doesn't change anything open source, the risk of this is low.

Differential Revision: [D17558142](https://our.internmc.facebook.com/intern/diff/D17558142/)

smessmer committed:
Switch internal CUDA build to C++14

This doesn't switch any open source builds or CI.
The internal fbcode build is C++17 already for quite some time, but in CUDA code, we had it restricted to C++11.
This diff changes that to C++14.

Because this doesn't change anything open source, the risk of this is low.

Differential Revision: [D17558142](https://our.internmc.facebook.com/intern/diff/D17558142/)

[ghstack-poisoned]

smessmer committed:
Update on "Switch internal CUDA build to C++14"


This doesn't switch any open source builds or CI.
The internal fbcode build is C++17 already for quite some time, but in CUDA code, we had it restricted to C++11.
This diff changes that to C++14.

Because this doesn't change anything open source, the risk of this is low.

Differential Revision: [D17558142](https://our.internmc.facebook.com/intern/diff/D17558142/)

[ghstack-poisoned]

ezyang commented:
> Because this doesn't change anything open source, the risk of this is low.

But there are clearly changes in this exported diff! Is it just moving code around?

smessmer committed:
Update on "Switch internal CUDA build to C++14"


This doesn't switch any open source builds or CI.
The internal fbcode build is C++17 already for quite some time, but in CUDA code, we had it restricted to C++11.
This diff changes that to C++14.

Because this doesn't change anything open source, the risk of this is low.

Differential Revision: [D17558142](https://our.internmc.facebook.com/intern/diff/D17558142/)

[ghstack-poisoned]

smessmer commented:
> > Because this doesn't change anything open source, the risk of this is low.
> 
> But there are clearly changes in this exported diff! Is it just moving code around?

Enabling C++14 in internal CUDA requires us to work around some CUDA bugs, and these are the changes you see here. What I meant is that it doesn't switch our oss builds to C++14.
